packages/zambdas/src/ehr/get-unsolicited-results-resources/helpers.ts

```diff
@@ -1,22 +1,174 @@
+        if (isPatientSubject) {
+          const patientRef = resource.subject?.reference;
+          if (patientRef) patientRefToRelatedDrMap[patientRef] = resource.id;
+        }
+      }
+    }
+    if (resource.resourceType === 'Task' && resource.status === 'ready') readyTasks.push(resource);
+    if (resource.resourceType === 'Patient') patients.push(resource);
+  });
+  readyTasks.forEach((task) => {
+    const relatedDrId = task.basedOn
+      ?.find((ref) => ref.reference?.startsWith('DiagnosticReport'))
+      ?.reference?.replace('DiagnosticReport/', '');
+
+    if (relatedDrId) {
+      drMap[relatedDrId].readyTasks.push(task);
+    }
+  });
+  patients.forEach((patient) => {
+    const patientRef = `Patient/${patient.id}`;
+    const drId = patientRefToRelatedDrMap[patientRef];
+    drMap[drId].patient = patient;
+  });
+  return drMap;
+};
+
+const formatResourcesForResponse = (readyTasksByDr: ReadyTasksByDr): UnsolicitedResultTaskRowDTO[] => {
+  const rows = Object.values(readyTasksByDr).flatMap((relatedResources) => {
+    const taskDetails: UnsolicitedResultTaskRowDTO[] = [];
+    relatedResources.readyTasks.forEach((task) => {
+      const taskCode = task.code?.coding?.find((c) => c.system === LAB_ORDER_TASK.system)?.code;
+      if (taskCode && taskIsLabRelated(taskCode)) {
+        const { diagnosticReport, patient } = relatedResources;
+        const taskRowDescription = getURDescriptionText(taskCode, diagnosticReport, patient);
+        const actionText = getURActionText(taskCode);
+
+        const row: UnsolicitedResultTaskRowDTO = {
+          diagnosticReportId: diagnosticReport.id || 'unknown',
+          actionText,
+          taskRowDescription,
+          resultsReceivedDateTime: task.authoredOn || 'unknown',
+        };
+        taskDetails.push(row);
+      }
+    });
+    return taskDetails;
+  });
+  return rows;
+};
+
+const getURDescriptionText = (code: string, diagnosticReport: DiagnosticReport, patient?: Patient): string => {
+  // todo sarah lab name should be included in the task row instruction
+  // currently this is blocked by some oystehr dev to be able to grab the lab off of DR
+  const testName = getTestNameFromDR(diagnosticReport);
+  const testItemCode = getTestItemCodeFromDR(diagnosticReport);
+  const testDescription = testName || testItemCode || '';
+
+  // only matched results will have patient linked
+  const patientName = patient ? getFullestAvailableName(patient, true) : undefined;
+
+  switch (code) {
+    case LAB_ORDER_TASK.code.matchUnsolicitedResult: {
+      return 'Match unsolicited test results';
+    }
+    case LAB_ORDER_TASK.code.reviewCancelledResult:
+    case LAB_ORDER_TASK.code.reviewCorrectedResult:
+    case LAB_ORDER_TASK.code.reviewFinalResult:
+    case LAB_ORDER_TASK.code.reviewPreliminaryResult: {
+      if (!patientName) throw Error('Cannot parse patient name for a matched unsolicited result');
```

Comment:
Probs worth throwing the DR id or task id or something in the error for debugging. Also slightest of nits on the wording of the error
```suggestion
      if (!patientName) throw Error('Cannot parse patient name for this matched unsolicited result');
```